### PR TITLE
Refactors ConversationsProvider.tsx

### PR DESCRIPTION
**Pull request by rhonsby**
### Description

I'm in the process of adding multiple inboxes, but I noticed that `ConversationsProvider.tsx` is quite bloated and has some properties that aren't even used anymore. This is the first set of changes to refactor this file as I didn't want this commit to get too large.  

I've broken up each set of changes into specific commits so it might be easier to look at the changes by the commits.

### Screenshots

No frontend changes, but did verify that UI still works as intended. 

## Checklist

- [x] Everything passes when running `mix test`
- [x] Ran `mix format`
- [x] No frontend compilation warnings

### CODE REVIEW

File: assets/src/components/conversations/ConversationNotificationManager.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ * connecting the React app with the Phoenix server for real-time messaging.
+ */
+class ConversationNotificationManager {
+  config: Config;
+  socket: Socket;
+  channel: Channel | null = null;
+
+  constructor(config: Config) {
+    this.socket = this.createNewSocket();
+    this.config = config;
```
Comment (rhonsby) [inline assets/src/components/conversations/ConversationNotificationManager.ts:30]: I kept most of the logic the same and used the `config` constructor argument to pass in the callbacks.

File: assets/src/components/conversations/ConversationsProvider.tsx [hunk 2]
```diff
@@ -207,106 +126,29 @@ export class ConversationsProvider extends React.Component<Props, State> {
...
-      if (!combined[key]) {
-        return acc;
-      }
-
-      return {...acc, [key]: combined[key]};
-    }, {} as PhoenixPresence);
-
-    this.setState({presence: latest});
+    this.setState({
+      presence: updatePresenceWithDiff(this.state.presence, diff),
```
Comment (rhonsby) [inline assets/src/components/conversations/ConversationsProvider.tsx:150]: By refactoring presence related logic into its own file, I was able to extract away everything and only expose this one method for diffing presence.

File: assets/src/presence.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
+type PresenceMetadata = {online_at?: string; phx_ref: string};
```
Comment (rhonsby) [inline assets/src/presence.ts:1]: All logic is the same as it was in `ConversationsProvider.tsx`. I just extracted them into this file and created an additional helper method, `updatePresenceWithDiff`, for diffing presence.

**Review by reichert621 — COMMENTED** [review]
ahhh this is amazing @rhonsby ❤️ thanks!

i mostly just skimmed through it but everything seems to make sense 👍 (i think i'll just want to do a bit of manual QA before merging since the `ConversationsProvider` manages a lot of the core functionality of the dashboard chat)

**Comment by rhonsby** [comment]
@reichert621 let me know when you get a chance to QA this. Would love to merge it in today!

**Comment by reichert621** [comment]
sounds good @rhonsby! will get to this by today for sure :+1: in the meantime feel free to branch off of this branch because i'm assuming any feedback will be minor (if any)

File: assets/src/components/conversations/ConversationsProvider.tsx [hunk 4]
```diff
@@ -63,98 +63,18 @@ export const ConversationsContext = React.createContext<{
...
-  fetchPriorityConversations: () => Promise.resolve([]),
-  fetchClosedConversations: () => Promise.resolve([]),
   fetchConversationById: () => Promise.resolve([]),
 });
 
 export const useConversations = () => useContext(ConversationsContext);
 
 type ConversationBucket = 'all' | 'mine' | 'priority' | 'closed';
 
-type PresenceMetadata = {online_at?: string; phx_ref: string};
```
Comment (reichert621) [inline assets/src/components/conversations/ConversationsProvider.tsx:82]: 🔥 🔥 🔥

**Review by reichert621 — APPROVED** [review]
nice! 🎉 

(all your stuff was good, just noticed one small unrelated bug when testing it out, so fixed that as part of this PR too 👍 )

**Pull request merged**